Kotlin Conversions

hiqua opened the pull request.

iSoron commented on uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/EmptyWidgetView.kt:
```diff
@@ -0,0 +1,43 @@
+import org.isoron.uhabits.R
+
+class EmptyWidgetView(context: Context?) : HabitWidgetView(context) {
+    private var title: TextView? = null
```
Please make this `lateinit var` instead

iSoron commented on uhabits-android/src/main/java/org/isoron/uhabits/widgets/views/GraphWidgetView.kt:
```diff
@@ -0,0 +1,51 @@
+import org.isoron.uhabits.R
+
+class GraphWidgetView(context: Context?, val dataView: View) : HabitWidgetView(context) {
+    private var title: TextView? = null
```
Please make this `lateinit var` instead

iSoron reviewed (changes requested):
Thank you for the pull request, @hiqua. This is great!

I have reviewed about half of the files; I will review the remaining ones soon.

hiqua commented:
Ready for a second review. I'll merge the fixing commits with the corresponding others once we're through.

iSoron commented on uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/AboutTest.kt:
```diff
@@ -0,0 +1,51 @@
+    @Test
+    fun shouldDisplayAboutScreen() {
+        launchApp()
+        ListHabitsSteps.clickMenu(ListHabitsSteps.MenuItem.ABOUT)
```
Could you static import this method too?

iSoron commented on uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/HabitsTest.kt:
```diff
@@ -0,0 +1,192 @@
+        clickSave()
+        verifyShowsScreen(CommonSteps.Screen.SHOW_HABIT)
+        verifyDisplaysText("Take a walk")
+        CommonSteps.pressBack()
```
Could you static-import this method?

iSoron commented on uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/LinksTest.kt:
```diff
@@ -0,0 +1,68 @@
+    @Throws(Exception::class)
+    fun shouldLinkToSourceCode() {
+        launchApp()
+        ListHabitsSteps.clickMenu(ListHabitsSteps.MenuItem.ABOUT)
```
Could you static-import this method?

iSoron commented on uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/ListHabitsRootView.kt:
```diff
@@ -126,17 +124,17 @@ class ListHabitsRootView @Inject constructor(
         val count = getCheckmarkCount()
         header.buttonCount = count
-        header.setMaxDataOffset(max(MAX_CHECKMARK_COUNT - count, 0))
+        header.setMaxDataOffset(kotlin.math.max(MAX_CHECKMARK_COUNT - count, 0))
```
Could you static-import `max` and `min`?

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:
```diff
@@ -0,0 +1,159 @@
+
+@RunWith(MockitoJUnitRunner::class)
+open class BaseUnitTest {
+    protected open var habitList: HabitList? = null
+    protected var fixtures: HabitFixtures? = null
+    protected var modelFactory: ModelFactory? = null
+    protected var taskRunner: SingleThreadTaskRunner? = null
+    protected open var commandRunner: CommandRunner? = null
```
I think the majority of these are initialized in `setUp`, so they could be `lateinit var` instead

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:
```diff
@@ -0,0 +1,159 @@
+
+    fun unixTime(year: Int, month: Int, day: Int): Long {
+        val cal = getStartOfTodayCalendar()
+        cal[year, month, day, 0, 0] = 0
```
It seems like the automatic Kotlin conversion transforms `cal.set(year, month, day, hour, min, sec)` into `cal[year, month, day, hour, min] = sec`. Could you correct this?

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:
```diff
@@ -0,0 +1,159 @@
+
+    open fun unixTime(year: Int, month: Int, day: Int, hour: Int, minute: Int): Long {
+        val cal = getStartOfTodayCalendar()
+        cal[year, month, day, hour] = minute
```
Same here

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/BaseUnitTest.kt:
```diff
@@ -0,0 +1,159 @@
+
+    @Throws(IOException::class)
+    protected fun openAsset(assetPath: String): InputStream {
+        var `in` = javaClass.getResourceAsStream(assetPath)
+        if (`in` != null) return `in`
+        val pwd = Paths.get(".").toAbsolutePath().normalize().toString()
+        val fullPath = "$pwd/assets/test/$assetPath"
+        val file = File(fullPath)
+        if (file.exists() && file.canRead()) `in` = FileInputStream(file)
+        if (`in` != null) return `in`
+        throw IllegalStateException("asset not found: $fullPath")
```
Could you rename `in` so we don't need backticks?

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/ChangeHabitColorCommandTest.kt:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private fun checkNewColors() {
+        for ((color) in selected!!) MatcherAssert.assertThat(
```
Please replace `(color) in selected!!` by `habit in selected!!`, and then `color` by `habit.color`.

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/ChangeHabitColorCommandTest.kt:
```diff
@@ -0,0 +1,68 @@
+
+    private fun checkOriginalColors() {
+        var k = 0
+        for ((color) in selected!!) MatcherAssert.assertThat(
```
Same here

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/CreateHabitCommandTest.kt:
```diff
@@ -0,0 +1,57 @@
+        Assert.assertTrue(habitList!!.isEmpty)
+        command!!.run()
+        MatcherAssert.assertThat(habitList!!.size(), CoreMatchers.equalTo(1))
+        val (_, _, _, _, _, name) = habitList!!.getByPosition(0)
```
Please replace this by `val habit = habitList!!.getByPosition(0)`

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/EditHabitCommandTest.kt:
```diff
@@ -0,0 +1,61 @@
+import org.junit.Test
+
+class EditHabitCommandTest : BaseUnitTest() {
+    private var command: EditHabitCommand? = null
+    private var habit: Habit? = null
+    private var modified: Habit? = null
+    private var today: Timestamp? = null
```
Please replace these by `lateinit var`

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/EditHabitCommandTest.kt:
```diff
@@ -0,0 +1,61 @@
+    fun testExecute() {
+        command = EditHabitCommand(habitList!!, habit!!.id!!, modified!!)
+        val originalScore = habit!!.scores[today!!].value
+        MatcherAssert.assertThat(habit!!.name, Matchers.equalTo("original"))
```
Could you static-import `assertThat` and `equalTo`?

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/commands/UnarchiveHabitsCommandTest.kt:
```diff
@@ -16,43 +16,31 @@
-        command.run();
-        assertTrue(habit.isArchived());
+    fun testExecuteUndoRedo() {
+        Assert.assertTrue(habit.isArchived)
+        command!!.run()
+        Assert.assertFalse(habit.isArchived)
```
Please static-import `assert` methods

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/database/migrations/Version22Test.kt:
```diff
@@ -0,0 +1,219 @@
+class Version22Test : BaseUnitTest() {
+    @get:Rule
+    var exception = ExpectedException.none()!!
+    private var db: Database? = null
+    private var helper: MigrationHelper? = null
```
Please replace by `lateinit var`

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.kt:
```diff
@@ -0,0 +1,130 @@
+        habitList!!.add(fixtures!!.createShortHabit())
+        habitList!!.add(fixtures!!.createEmptyHabit())
+        baseDir = Files.createTempDirectory("csv").toFile()
+        Assert.assertNotNull(baseDir)
```
Please static-import assert methods

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.kt:
```diff
@@ -0,0 +1,130 @@
+        zip.close()
+    }
+
+    //    @Test
+    //    public void test_writeCSV() throws IOException
+    //    {
+    //        Habit habit = fixtures.createShortHabit();
+    //
+    //        String expectedCSV =
+    //                "2015-01-25,0.2557\n" +
+    //                        "2015-01-24,0.2226\n" +
+    //                        "2015-01-23,0.1991\n" +
+    //                        "2015-01-22,0.1746\n" +
+    //                        "2015-01-21,0.1379\n" +
+    //                        "2015-01-20,0.0995\n" +
+    //                        "2015-01-19,0.0706\n" +
+    //                        "2015-01-18,0.0515\n" +
+    //                        "2015-01-17,0.0315\n" +
+    //                        "2015-01-16,0.0107\n";
+    //
+    //        StringWriter writer = new StringWriter();
+    //        habit.getScores().writeCSV(writer);
+    //
+    //        assertThat(writer.toString(), equalTo(expectedCSV));
+    //    }
```
Please delete this. Also, feel free to delete any other commented-out tests.

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/ImportTest.kt:
```diff
@@ -0,0 +1,138 @@
+    @Test
+    @Throws(IOException::class)
+    fun testLoopDB() {
+        importFromFile("loop.db")
+        MatcherAssert.assertThat(habitList!!.size(), IsEqual.equalTo(9))
+        val habit = habitList!!.getByPosition(0)
+        MatcherAssert.assertThat(habit.name, IsEqual.equalTo("Wake up early"))
+        MatcherAssert.assertThat(habit.frequency, IsEqual.equalTo(Frequency.THREE_TIMES_PER_WEEK))
+        Assert.assertTrue(isChecked(habit, 2016, 3, 14))
+        Assert.assertTrue(isChecked(habit, 2016, 3, 16))
+        TestCase.assertFalse(isChecked(habit, 2016, 3, 17))
```
Please static-import assert methods

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/io/ImportTest.kt:
```diff
@@ -0,0 +1,138 @@
+
+    private fun isChecked(h: Habit, year: Int, month: Int, day: Int): Boolean {
+        val date = getStartOfTodayCalendar()
+        date[year, month - 1] = day
```
Please replace this by `date.set(year, month-1, day)`

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:
```diff
@@ -0,0 +1,307 @@
+class HabitListTest : BaseUnitTest() {
+    @get:Rule
+    var thrown = ExpectedException.none()!!
+    private var habitsArray: ArrayList<Habit>? = null
+    private var activeHabits: HabitList? = null
+    private var reminderHabits: HabitList? = null
```
Please replace by `lateinit var`

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:
```diff
@@ -0,0 +1,307 @@
+
+    @Test
+    fun testSize() {
+        MatcherAssert.assertThat(habitList!!.size(), CoreMatchers.equalTo(10))
+        MatcherAssert.assertThat(activeHabits!!.size(), CoreMatchers.equalTo(6))
+        MatcherAssert.assertThat(reminderHabits!!.size(), CoreMatchers.equalTo(4))
```
Please use static imports in `assertThat` and `equalTo`. Same for the other usages in this file.

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:
```diff
@@ -0,0 +1,307 @@
+
+    @Test
+    fun testOrdering() {
+        val h1 = fixtures!!.createEmptyHabit()
+        h1.name = "A Habit"
+        h1.color = PaletteColor(2)
+        h1.position = 1
```
Could you use `apply` to make this look nicer?
```kotlin
val h1 = fixtures.createEmptyHabit().apply {
    name = "A Habit"
    color = [...]
}
```
An alternative would be making `name`, `color`, `position` etc arguments of `createEmptyHabit`, but provide default values.

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:
```diff
@@ -0,0 +1,307 @@
+        h4.name = "D Habit"
+        h4.color = PaletteColor(1)
+        h4.position = 2
+        val list = modelFactory!!.buildHabitList()
+        list.add(h3)
+        list.add(h1)
+        list.add(h4)
+        list.add(h2)
```
`apply` could also be used here

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:
```diff
@@ -0,0 +1,307 @@
+            habitList!!.reorder(fromHabit, toHabit)
+            val actualSequence = IntArray(10)
+            for (j in 0..9) {
+                val (_, _, _, id, _, _, position) = habitList!!.getByPosition(j)
```
Please change to `val habit = habitList!!.getByPosition(j)`

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/models/HabitListTest.kt:
```diff
@@ -0,0 +1,301 @@
+
+    @Test
+    fun testGetByPosition() {
+        assertThat(
+            habitList.getByPosition(0),
+            CoreMatchers.equalTo(
+                habitsArray[0]
+            )
+        )
```
Please put these in a single line. I think ktlint breaks lines, but doesn't merge them back.

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/ui/screens/habits/list/ListHabitsMenuBehaviorTest.kt:
```diff
@@ -0,0 +1,214 @@
+        verify(adapter).setSecondaryOrder(
+            secondaryOrderCaptor.capture()
+        )
+        assertThat(
+            orderCaptor.lastValue,
+            equalTo(HabitList.Order.BY_STATUS_ASC)
+        )
```
I think these could be in a single line. There are many other similar cases in this file.

iSoron commented on uhabits-core/src/jvmTest/java/org/isoron/uhabits/core/utils/DateUtilsTest.kt:
```diff
@@ -0,0 +1,479 @@
+        var t0 = unixTime(2015, Calendar.JANUARY, 11)
+        var t1 = unixTime(2015, Calendar.JANUARY, 16)
+        var t2 = unixTime(2015, Calendar.JANUARY, 17)
+        assertThat(truncate(field, t0, firstWeekday), CoreMatchers.equalTo(expected))
+        assertThat(truncate(field, t1, firstWeekday), CoreMatchers.equalTo(expected))
+        assertThat(truncate(field, t2, firstWeekday), CoreMatchers.equalTo(expected))
```
Please use static imports

iSoron reviewed (changes requested):
Just a few minor changes requested. Hopefully we can merge it after these are fixed.

hiqua commented:
I've removed the assertNonNull and replaced them with `!!`. These are bound to disappear completely anyway, once we've got rid of (most of) the nullable types in kotlin.

hiqua commented:
@iSoron can you update https://github.com/iSoron/uhabits/issues/685 once we're done with this PR?

iSoron commented:
Merged! Thanks for this PR, @hiqua. I'm sure it was a lot of work. I'll update #685 soon.
